### PR TITLE
clientv3: Health balancer

### LINKED ISSUE #8326: Automatically switch endpoint when one endpoint is not available
we found when 1 of 3 instance of etcd is network partition witch other 2, 

still the client can access this instance, and grpc connection is keep alive, then all request will failed and report like this

```
#kubectl get node --all-namespaces
Error from server: rpc error: code = 14 desc = etcdserver: request time out
```

out etcd version is 3.1.9

this error is not happen is etcd v2

### PR DESCRIPTION
**Pull request by heyitsanthony**
Works with the watch keepalive test from the grpc ping timeout patch.

Supersedes #8327
Fixes #8326

### CODE REVIEW

File: clientv3/retry.go [hunk 1]
```diff
@@ -46,7 +46,14 @@ func isWriteStopError(err error) bool {
 func (c *Client) newRetryWrapper(isStop retryStopErrFunc) retryRpcFunc {
 	return func(rpcCtx context.Context, f rpcFunc) error {
 		for {
-			if err := f(rpcCtx); err == nil || isStop(err) {
+			err := f(rpcCtx)
+			if err == nil {
+				return nil
+			}
+			if grpc.Code(err) == codes.Unavailable {
```
Comment (gyuho) [inline clientv3/retry.go]: gRPC is deprecating `grpc.Code`.

```go
ev, _ := status.FromError(err)
if ev.Code() == codes.Unavailable {
```

?

File: clientv3/health_balancer.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+		return false, err
+	}
+	defer conn.Close()
+	cli := healthpb.NewHealthClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	resp, err := cli.Check(ctx, &healthpb.HealthCheckRequest{})
+	cancel()
+	if err != nil {
+		if ev, _ := status.FromError(err); ev.Code() == codes.Unavailable &&
+			grpc.ErrorDesc(err) == "unknown service grpc.health.v1.Health" {
```
Comment (xiang90) [inline clientv3/health_balancer.go]: define this string as const?

**Comment by xiang90** [comment]
lgtm in general. @gyuho or @fanminshi should give this a more close look.

File: clientv3/health_balancer.go (hunk 2 above)
Comment (gyuho) [reply clientv3/health_balancer.go]: `s/grpc.ErrorDesc/rpctypes.ErrorDesc`?

`grpc.ErrorDesc` is being deprecated.

File: clientv3/health_balancer.go [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+}
+
+func (hb *healthBalancer) liveAddrs() []grpc.Address {
+	hb.mu.RLock()
+	defer hb.mu.RUnlock()
+	hbAddrs := hb.addrs
+	if len(hb.addrs) == 1 || len(hb.unhealthy) == 0 || len(hb.unhealthy) == len(hb.addrs) {
+		return hbAddrs
+	}
+	addrs := make([]grpc.Address, len(hb.addrs)-len(hb.unhealthy))
```
Comment (gyuho) [inline clientv3/health_balancer.go]: `s/make([]grpc.Address, len(hb.addrs)-len(hb.unhealthy))/make([]grpc.Address, 0, len(hb.addrs)-len(hb.unhealthy))`

**Comment by heyitsanthony** [comment]
got this working; `next` needed to wait until the connection went down so the retried RPC wouldn't race with the balancer and use the old connection

File: clientv3/balancer.go [hunk 4]
```diff
@@ -135,27 +170,37 @@ func (b *simpleBalancer) updateAddrs(eps []string) {
...
 }
 
+func (b *simpleBalancer) next() {
+	b.mu.RLock()
+	downc := b.downc
+	b.mu.RUnlock()
+	select {
+	case b.updateAddrsC <- notifyNext:
+	case <-b.stopc:
+	}
```
Comment (gyuho) [inline clientv3/balancer.go]: Should we return on `<-b.stopc`?

Comment (heyitsanthony) [reply clientv3/balancer.go]: the channel is closed, so it'll pass through the next select without blocking

**Pull request merged**